bodo/io/arrow_reader.h

```diff
@@ -543,6 +545,40 @@ class ArrowReader {
     bool reported_init_stage_metrics = false;
     bool reported_read_stage_metrics = false;
 
+    const int32_t batch_readahead = init_batch_readahead();
+    const int32_t frag_readahead = init_frag_readahead();
+
+    /*
+     * @brief determine how many batches to read ahead. Arrows defaults are
+     * based on a single reader and we need to adjust for the number of ranks on
+     * the node while always ensuring at least one batch is read ahead.
+     */
+    int32_t init_batch_readahead() {
+        if (!this->parallel) {
+            // If not parallel, we don't need to adjust the readahead
+            return arrow::dataset::kDefaultBatchReadahead;
+        }
+        int ranksOnNode = std::get<0>(dist_get_ranks_on_node());
+        return std::max(arrow::dataset::kDefaultBatchReadahead / ranksOnNode,
+                        1);
+    }
+
+    /*
+     * @brief determine how many batches to read ahead. Arrows defaults are
+     * based on a single reader and we need to adjust for the number of ranks on
+     * the node. We don't make sure this is at least one because fragments can
+     * be large, and if we have a lot of ranks a full extra fragment can be too
+     * much memory.
+     */
+    int32_t init_frag_readahead() {
+        if (!this->parallel) {
+            // If not parallel, we don't need to adjust the readahead
+            return arrow::dataset::kDefaultFragmentReadahead;
+        }
+        int ranksOnNode = std::get<0>(dist_get_ranks_on_node());
+        return arrow::dataset::kDefaultFragmentReadahead / ranksOnNode;
```

Comment:
This looks a bit scary in terms of performance. This will be 0 on any machine with more than 4 cores? Is 1 a better minimum? Have you measured performance vs main? Is there a way to be more adaptive and adjust based on fragment size?

Reply:
Ok, thanks for testing.